### PR TITLE
lockstep refactoring of `KernelDeriveParticles`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- rewrite kernel with the lockstep programming model
- add documentation
- renaming of some variables and types

[HowTo lockstep programming](http://picongpu.readthedocs.io/en/dev/prgpatterns/lockstep.html)

# Tests

- [x] default 3D KHI

### CODE REVIEW

File: src/picongpu/include/particles/Particles.kernel [hunk 1]
```diff
@@ -50,69 +46,155 @@
...
 namespace picongpu
 {
 
-using namespace PMacc;
-
-
+/** derive new particles from a source species
+ *
+ * This functor prepare a source and destination particle box to call
+ * a user defined functor which allows o derive new particles out of
```
Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: `o derive` sounds Portuguese :) -> *to derive*

File: src/picongpu/include/particles/Particles.kernel [hunk 2]
```diff
@@ -50,69 +46,155 @@
...
 
 namespace picongpu
 {
 
-using namespace PMacc;
-
-
+/** derive new particles from a source species
+ *
+ * This functor prepare a source and destination particle box to call
```
Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: prepare**s**

File: src/picongpu/include/particles/Particles.kernel [hunk 3]
```diff
@@ -50,69 +46,155 @@
...
+ * another species.
+ *
+ * @tparam T_numWorkers number of workers
+ */
+template< uint32_t T_numWorkers >
 struct KernelDeriveParticles
 {
-    template<class T_MyParBox, class T_OtherFrameBox, class T_ManipulateFunctor, class Mapping>
-    DINLINE void operator()(T_MyParBox myBox, T_OtherFrameBox otherBox, T_ManipulateFunctor manipulateFunctor, Mapping mapper) const
+    /** frame wise derive new particles
```
Comment (ax3l) [inline src/picongpu/include/particles/Particles.kernel]: maybe written *frame-wise*

**Comment by psychocoderHPC** [comment]
updated

**Pull request merged**